flow/Histogram.h

```diff
@@ -35,64 +35,80 @@
 };
 
-// TODO: This should be scoped properly for simulation (instead of just having all the "machines" share one histogram namespace)
-HistogramRegistry & GetHistogramRegistry();
+// TODO: This should be scoped properly for simulation (instead of just having all the "machines" share one histogram
+// namespace)
+HistogramRegistry& GetHistogramRegistry();
 
-class Histogram {
+class Histogram : public ReferenceCounted<Histogram> {
 public:
-    enum class Unit {
-        microseconds,
-        bytes
-    };
+	enum class Unit { microseconds, bytes };
 
-    Histogram(StringRef group, StringRef op, Unit unit) : group(group.toString()), op(op.toString()), unit(unit), registry(GetHistogramRegistry()) {
-        clear();
-        registry.registerHistogram(this);
-    }
+private:
+	Histogram(std::string group, std::string op, Unit unit)
+	  : group(group), op(op), unit(unit), registry(GetHistogramRegistry()) {
+		clear();
+		registry.registerHistogram(this);
+	}
 
-    ~Histogram() {
-        registry.unregisterHistogram(this);
-    }
+	static std::string generateName(std::string group, std::string op) { return group + ":" + op; }
 
-    inline void sample(uint32_t sample) {
+public:
+	~Histogram() { registry.unregisterHistogram(this); }
+
+	static Reference<Histogram> getHistogram(StringRef group, StringRef op, Unit unit) {
+		std::string group_str = group.toString();
+		std::string op_str = op.toString();
+		std::string name = generateName(group_str, op_str);
+		HistogramRegistry& registry = GetHistogramRegistry();
+		Histogram* h = registry.lookupHistogram(name);
+		if (!h) {
+			h = new Histogram(group_str, op_str, unit);
+		}
+		return Reference(h);
+	}
+
+	inline void sample(uint32_t sample) {
 #ifdef _WIN32
-        unsigned long index;
-        buckets[_BitScanReverse(&index, sample) ? index : 0]++;
+		unsigned long index;
+		buckets[_BitScanReverse(&index, sample) ? index : 0]++;
 #else
-        buckets[sample ? (31 - __builtin_clz(sample)) : 0]++;
+		buckets[sample ? (31 - __builtin_clz(sample)) : 0]++;
 #endif
-    }
-
-    inline void sampleSeconds(double delta) {
-        sample((uint32_t)(delta * 1000000)); // convert to microseconds and truncate to integer
-    }
-
-    void clear() {
-        for (uint32_t & i : buckets) {
-            i = 0;
-        }
-    }
-    void writeToLog();
-
-    std::string name() {
-        return group + ":" + op;
-    }
-
-    std::string const group;
-    std::string const op;
-    Unit const unit;
-    HistogramRegistry & registry;
-    
-private:
-    uint32_t buckets[32];
+	}
+
+	inline void sampleSeconds(double delta) {
+		uint64_t delta_usec = (delta * 1000000);
+		if (delta_usec > UINT32_MAX) {
+			sample(UINT32_MAX);
+		} else {
+			sample((uint32_t)(delta * 1000000)); // convert to microseconds and truncate to integer
+		}
+	}
+
+	void clear() {
+		for (uint32_t& i : buckets) {
+			i = 0;
+		}
+	}
+	void writeToLog();
+
+	std::string name() { return generateName(this->group, this->op); }
```

Comment:
I am wondering if we could just

```C++
const std::string name;
```

and initialize it in the constructor?

Reply:
I went back and forth on that.  This runs once per histogram every report interval (currently 5 seconds, but destined to be 30sec - 5min), so the performance difference is negligible.